fix(core): ICU expression previous value not removed inside template

crisbeto opened the pull request:
Fixes an issue where the previous value of an ICU expression wasn't being
removed, if it didn't have an `other` parameter and it's the last value inside
a template. The issue is due to the `createDynamicNodeAtIndex` function
which sets a new node at a particular index inside the `LView`, but it doesn't
account for any node that may have existed there beforehand. Since we've
overwritten the reference, we don't have a way of removing the node from
the DOM anymore.

Also cleans up an optional parameter in one of the internal functions and
adds stronger typing to the i18n unit tests.

Fixes #38073.

petebacondarwin reviewed (commented):
You absolute legend @crisbeto !!

AndrewKushnir commented on packages/core/test/acceptance/i18n_spec.ts:
```diff
@@ -1435,6 +1435,19 @@ onlyInIvy('Ivy i18n logic').describe('runtime i18n', () => {
+      fixture.detectChanges();
+      expect(fixture.nativeElement.textContent.trim()).toBe('AB');
+
+      fixture.componentInstance.items[0].value = 2;
```
I think it'd be great to add a couple more tests/checks that would:
- update the value of an element that was not rendered originally (i.e. `items[2].value` in this case) and set the value to one of the conditions present in ICU
- change the value of an element that correspond to one of the cases in ICU (e.g. '1') to a value that would not be matched by ICU conditions (e.g. '5')

May be also copy all these tests for a standalone ICU (i.e. outside of an `ngFor`).

crisbeto replied:
I've added the extra test cases. As for copying them to be without an `ngFor`, I didn't do it because I wasn't able to reproduce the issue without it. I think it has more to do with the number of nodes than the template definition itself.

AndrewKushnir reviewed (commented):
Thanks for the fix @crisbeto 👍 

I just left a quick comment (with the proposal to check more cases) and will have closer look at the fix on Monday.

petebacondarwin commented on packages/core/src/render3/i18n.ts:
```diff
@@ -874,7 +874,7 @@ function readCreateOpCodes(
 function readUpdateOpCodes(
     updateOpCodes: I18nUpdateOpCodes, icus: TIcu[]|null, bindingsStartIndex: number,
-    changeMask: number, tView: TView, lView: LView, bypassCheckBit = false) {
+    changeMask: number, tView: TView, lView: LView, bypassCheckBit: boolean) {
```
This change is just a refactoring, yes?

petebacondarwin replied:
I read in the commit message that it is. (NIT: personally I would have made these refactoring type changes in another commit to make it really clear that it is not a required part of the fix).

crisbeto replied:
I figured that it was small enough that it wouldn't influence the review too much.

petebacondarwin reviewed (commented):
I have to admit I don't really understand the sequence of calls to `readUpdateOpCode()` but it seems strange that we have to leave the `activeCaseIndex` as non-null (at line 944), even though (from my reading of the code) it should have been cleared up earlier on in the function (lines 916-941).

petebacondarwin commented on packages/core/test/acceptance/i18n_spec.ts:
```diff
@@ -1435,6 +1435,45 @@ onlyInIvy('Ivy i18n logic').describe('runtime i18n', () => {
+      expect(fixture.nativeElement.textContent.trim()).toBe('CB');
+    });
+
+    it('should render an element whose case did no match initially', () => {
```
```suggestion
    it('should render an element whose case did not match initially', () => {
```

crisbeto commented:
My understanding of the entire process is somewhat high level, but I think the way it's set up is that it'll remove the old node next time it runs through the entire routine. If we reset it back to null, next time around it won't be able to remove the node since it doesn't know the index from last time.

petebacondarwin commented:
From what I can see, all the iterations of the `ngFor` loop use the same `icuTNode`, which means that on each loop the following happens:

1. remove the previous iteration's active case
2. create the current iteration's active case
3. update the current iteration's active case

This seems wrong to me... if the `icuTNode` is shared across all the ngFor iterations, then I would think it is not the right place to store the `activeCaseIndex`, right?

And I think it is this weird sharing that is causing the first iteration to be updated incorrectly, since the `activeCaseIndex`, from the last iteration, in the previous loop, is `null` it doesn't remove its previous active case before adding the new active case.

crisbeto commented:
I see what you mean now. Setting this back to draft and I'll take a better look later.

crisbeto committed:
fix(core): ICU expression previous value not removed inside template

Fixes an issue where the previous value of an ICU expression wasn't being
removed, if it didn't have an `other` parameter and it's the last value inside
a template. The issue is due to the `createDynamicNodeAtIndex` function
which sets a new node at a particular index inside the `LView`, but it doesn't
account for any node that may have existed there beforehand. Since we've
overwritten the reference, we don't have a way of removing the node from
the DOM anymore.

Also cleans up an optional parameter in one of the internal functions and
adds stronger typing to the i18n unit tests.

Fixes #38073.

AndrewKushnir commented:
Hi @crisbeto, it looks like https://github.com/angular/angular/commit/6d8c73a4d62380bd7cfcaeaa23d5db802467d485 supersedes the changes in this PR? If that's the case, can we close this PR? Thank you.
